File: src/swift/2_CreationalPatterns/4_Singleton/Exercise.swift
import Foundation
import XCTest

class SingletonTester
{
  static func isSingleton(factory: () -> AnyObject) -> Bool
  {
    let obj1 = factory()
    let obj2 = factory()
    return obj1 === obj2
  }
}

class UMBaseTestCase : XCTestCase {}

//@testable import Test

class Evaluate: UMBaseTestCase
{
  func simpleTest()
  {
    let obj = SingletonTester() // hehe, inception!
    XCTAssert(type(of: obj).isSingleton{obj})
    XCTAssertFalse(type(of: obj).isSingleton{ SingletonTester() })
  }
}

extension Evaluate
{
  static var allTests : [(String, (Evaluate) -> () throws -> Void)]
  {
    return [
      ("simpleTest", simpleTest)
    ]
  }
}

func main()
{
  XCTMain([testCase(Evaluate.allTests)])
}

main()

File: src/python/2_CreationalPatterns/4_Singleton/exercise.py
from unittest import TestCase
from copy import deepcopy

def is_singleton(factory):
    x = factory()
    y = factory()
    return x is y

class Evaluate(TestCase):
    def test_exercise(self):
        obj = [1, 2, 3]
        self.assertTrue(is_singleton(lambda: obj))
        self.assertFalse(is_singleton(lambda: deepcopy(obj)))